components/ws-manager-api/go/crd/v1/workspace_types.go

```diff
@@ -0,0 +1,226 @@
+// Copyright (c) 2022 Gitpod GmbH. All rights reserved.
+// Licensed under the GNU Affero General Public License (AGPL).
+// See License-AGPL.txt in the project root for license information.
+
+package v1
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// EDIT THIS FILE!  THIS IS SCAFFOLDING FOR YOU TO OWN!
+// NOTE: json tags are required.  Any new fields you add must have json tags for the fields to be serialized.
+
+// WorkspaceSpec defines the desired state of Workspace
+type WorkspaceSpec struct {
+	// +kubebuilder:validation:Required
+	Ownership Ownership `json:"ownership"`
+
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:Enum:=Regular;Prebuild;ImageBuild
+	Type WorkspaceType `json:"type"`
+
+	Class string `json:"class"`
+
+	// +kubebuilder:validation:Required
+	Image WorkspaceImages `json:"image"`
+
+	Initializer []byte `json:"initializer"`
+
+	Envvars []corev1.EnvVar `json:"envvars,omitempty"`
+
+	// +kubebuilder:validation:Required
+	WorkspaceLocation string `json:"workspaceLocation"`
+
+	Git *GitSpec `json:"git,omitempty"`
+
+	Timeout TimeoutSpec `json:"timeout"`
+
+	Admission AdmissionSpec `json:"admission"`
+
+	Ports []PortSpec `json:"ports,omitempty"`
+}
+
+type Ownership struct {
+	// +kubebuilder:validation:Required
+	Owner string `json:"owner"`
+	// +kubebuilder:validation:Required
+	WorkspaceID string `json:"workspaceID"`
+	Team        string `json:"team,omitempty"`
+	Tenant      string `json:"tenant,omitempty"`
+}
+
+type WorkspaceType string
+
+const (
+	WorkspaceTypeRegular    WorkspaceType = "Regular"
+	WorkspaceTypePrebuild   WorkspaceType = "Prebuild"
+	WorkspaceTypeImageBuild WorkspaceType = "ImageBuild"
+)
+
+type WorkspaceImages struct {
+	Workspace WorkspaceImage `json:"workspace"`
+	IDE       IDEImages      `json:"ide"`
+}
+
+type WorkspaceImage struct {
+	Ref *string `json:"ref,omitempty"`
+}
+
+type IDEImages struct {
+	Web        string `json:"web"`
+	Desktop    string `json:"desktop"`
+	Supervisor string `json:"supervisor"`
+}
+
+type GitSpec struct {
+	Username string `json:"username"`
+	Email    string `json:"email"`
+}
+
+type TimeoutSpec struct {
+	Time *metav1.Duration `json:"time,omitempty"`
+}
+
+type AdmissionSpec struct {
+	Level AdmissionLevel `json:"level"`
+}
+
+type AdmissionLevel string
+
+const (
+	AdmissionLevelOwner    AdmissionLevel = "owner"
+	AdmissionLevelEveryone AdmissionLevel = "everyone"
+)
+
+type PortSpec struct {
+	// +kubebuilder:validation:Required
+	Port uint32 `json:"port"`
+
+	// +kubebuilder:validation:Required
+	Visibility AdmissionLevel `json:"visibility"`
+}
+
+// WorkspaceStatus defines the observed state of Workspace
+type WorkspaceStatus struct {
+	// INSERT ADDITIONAL STATUS FIELD - define observed state of cluster
+	// Important: Run "make" to regenerate code after modifying this file
+
+	PodStarts  int    `json:"podStarts"`
```

Comment:
what does this mean?

Reply:
the number of pod restarts

Reply:
For e.g. implementing prebuilds on spot instances. Whenever we need to control workspace restarts on the workspace side.